source/common/router/config_impl.cc

```diff
@@ -85,11 +87,56 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost, const Json:
       rate_limit_policy_(route), shadow_policy_(route),
       priority_(ConfigUtility::parsePriority(route)) {
 
+  bool have_weighted_clusters = route.hasObject("weighted_clusters");
+  bool have_cluster = !cluster_name_.empty() || have_weighted_clusters;
+
   // Check to make sure that we are either a redirect route or we have a cluster.
-  if (!(isRedirect() ^ !cluster_name_.empty())) {
+  if (!(isRedirect() ^ have_cluster)) {
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
 
+  if (have_cluster) {
+    if (!cluster_name_.empty() && have_weighted_clusters) {
+      throw EnvoyException("routes must have either single or weighted_clusters as target");
+    }
+  }
+
+  // If this is a weighted_cluster, we create N internal route entries
+  // (called WeightedClusterEntry), such that each object is a simple
+  // single cluster, pointing back to the parent.
+  if (have_weighted_clusters) {
+    uint64_t total_weight = 0UL;
+
+    const Json::ObjectPtr weighted_clusters_json = route.getObject("weighted_clusters");
+
+    if (!weighted_clusters_json->hasObject("clusters")) {
+      throw EnvoyException("weighted_clusters specification has no clusters defined");
+    }
+
+    const std::vector<Json::ObjectPtr> cluster_list =
+        weighted_clusters_json->getObjectArray("clusters");
+    const std::string runtime_key_prefix =
+        weighted_clusters_json->getString("runtime_key_prefix", EMPTY_STRING);
+
+    if (cluster_list.empty()) {
```

Comment:
On second thought, I would keep this one for now, since we don't have yet schemas and this will be pretty broken if this list is empty, but I would delete the other one. It doesn't matter that much to me if you keep/remove the tests for the above one since it should still throw an exception and the test should pass, so I would just leave the tests.